src/crab/cfg.hpp

```diff
@@ -333,14 +341,36 @@ class cfg_t final {
     size_t size() const { return static_cast<size_t>(std::distance(begin(), end())); }
 
     void simplify() {
-        merge_blocks();
-        remove_unreachable_blocks();
-        remove_useless_blocks();
-        // after removing useless blocks there can be opportunities to
-        // merge more blocks.
-        merge_blocks();
-        remove_joining_blocks();
-        merge_blocks();
+        std::unordered_set<label_t> worklist(this->label_begin(), this->label_end());
+        while (!worklist.empty()) {
+            label_t label = *worklist.begin();
+            worklist.erase(label);
+
+            basic_block_t& bb = get_node(label);
+            if (bb.in_degree() == 1 && get_parent(label).out_degree() == 1) {
+                continue;
+            }
+            while (bb.out_degree() == 1) {
+                basic_block_t& next_bb = get_child(label);
+                if (&next_bb == &bb || next_bb.in_degree() != 1) {
+                    break;
+                }
+                if (next_bb.label() == m_exit) {
+                    m_exit = label;
+                }
+
+                bb.move_back(next_bb);
+                bb -= next_bb;
+                std::vector<label_t> children = next_bb.m_next;
+                for (const label_t& next_next_label : children) {
+                    basic_block_t& next_next_bb = get_node(next_next_label);
+                    bb >> next_next_bb;
+                }
+                remove(next_bb.label());
+
+                worklist.erase(next_bb.label());
```

Comment:
This line is unsafe and can access freed memory, since line 369 calls m_blocks.erase(_label).
m_blocks is a std::unordered_map and https://www.cplusplus.com/reference/unordered_map/unordered_map/erase/ explains: 

> Only the iterators and references to the elements removed are invalidated.

Indeed on the std::unordered_map implementation I just tested with, the memory in next_bb.label() is freed by the time line 371 is executed and it accesses potentially corrupted memory.
To fix, above line 358 add:
```
std::string next_label = next_bb.label();
```
and then replace all subsequent uses of next_bb.label() with next_label.

Reply:
Oops.

I think it's better to perform the removal last, since it invalidates `next_bb` and not only the label.